### PR TITLE
Add repeat Spark function

### PR DESCRIPTION
**Pull request by zhli1142015**
Doc: https://spark.apache.org/docs/latest/api/sql/#repeat
Code:https://github.com/apache/spark/blob/257a7883f2150e037eb05f8c7a111184103ad9a1/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala#L1885
BTW, `repeat` is used for other fucntions for Presto, so register this with a different name.
https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/registration/ArrayFunctionsRegistration.cpp#L160

### CODE REVIEW

**Comment by zhli1142015** [comment]
cc @PHILO-HE  and @rui-mo , thanks.

File: velox/functions/sparksql/Register.cpp
Comment (PHILO-HE) [inline velox/functions/sparksql/Register.cpp:463]: I think we can also add the name "repeat" as function alias, as it is Spark's function name. Temporarily we call "string_repeat" in Gluten to avoid conflict issue with presto.

File: velox/docs/functions/spark/string.rst
Comment (PHILO-HE) [inline velox/docs/functions/spark/string.rst:144]: string_repeat -> repeat
And then document "string_repeat" separately to just indicate it's an alias for "repeat" function.

File: velox/docs/functions/spark/string.rst
Comment (rui-mo) [inline velox/docs/functions/spark/string.rst:151]: nit: function -> spark:function?

File: velox/docs/functions/spark/string.rst
Comment (rui-mo) [inline velox/docs/functions/spark/string.rst:149]: nit: SSELECT -> SELECT

File: velox/docs/functions/spark/string.rst
Comment (rui-mo) [inline velox/docs/functions/spark/string.rst:147]: Why do we add this limitation? I tried in Spark and found a larger value for n was allowed.

Comment (rui-mo) [reply velox/docs/functions/spark/string.rst:147]: I would be better to document the behavior when n is negative.

File: velox/functions/sparksql/tests/StringTest.cpp
Comment (rui-mo) [inline velox/functions/sparksql/tests/StringTest.cpp:462]: Can we add tests for negative times and empty input string?

**Review by rui-mo — COMMENTED** [review]
Thanks!

File: velox/functions/sparksql/tests/StringTest.cpp
Comment (zhli1142015) [reply velox/functions/sparksql/tests/StringTest.cpp:464]: Added, thanks

File: velox/docs/functions/spark/string.rst
Comment (zhli1142015) [reply velox/docs/functions/spark/string.rst]: Updated the doc.
Spark fuzzer test failed with exit code 137 which indicates out of memory error. I thought this is related to this parameter.
 I think 10000 times is big enough, not sure if larger vaule would be use in real cases. Or we should disable this fucntion in fuzzer test, please suggest.

File: velox/docs/functions/spark/string.rst
Comment (PHILO-HE) [inline velox/docs/functions/spark/string.rst:152]: Please also put this in alphabetical order, though it's just for documenting alias.

File: velox/docs/functions/spark/string.rst
Comment (PHILO-HE) [reply velox/docs/functions/spark/string.rst]: @zhli1142015, could you get the input size when OOM happens? Maybe, we should limit the value of `input_size * times`. If this value is greater than some threshold, let the execution fail with exception.

File: velox/docs/functions/spark/string.rst
Comment (PHILO-HE) [inline velox/docs/functions/spark/string.rst:148]: -> If \``n\`` is less than 0 or equal to 0,
Or
-> If \``n <= 0\``,

File: velox/docs/functions/spark/string.rst
Comment (PHILO-HE) [reply velox/docs/functions/spark/string.rst]: Or let's wait for the review from @mbasmanova.

Comment (PHILO-HE) [reply velox/docs/functions/spark/string.rst]: @zhli1142015, I see. The limit is an empirical value. I'm not sure how to handle. @mbasmanova, do you have any idea? Should we just let fuzzer test ignore the out-of-memory failure instead of introducing an empirical limit?

Comment (rui-mo) [reply velox/docs/functions/spark/string.rst]: I wonder if we can introduce a flag like `FLAGS_repeat_times` and set it as an acceptable value when running the fuzzer test to avoid OOM.

File: velox/docs/functions/spark/string.rst
Comment (rui-mo) [inline velox/docs/functions/spark/string.rst:144]: nit: use the full name string to replace str.

File: velox/docs/functions/spark/string.rst
Comment (rui-mo) [inline velox/docs/functions/spark/string.rst:146]: nit: "string value" could be changed as \``string`` to refer to the input variable.

File: velox/functions/sparksql/String.h
Comment (rui-mo) [inline velox/functions/sparksql/String.h:1203]: It seems the two branches could be combined as below.

> if (times <= 0 || input.size() == 0) {
result.resize(0);
return;
}

File: velox/functions/sparksql/String.h
Comment (rui-mo) [inline velox/functions/sparksql/String.h:1204]: In Spark, integer overflow exception is thrown if numBytes * times exceeds the max value of int. Could we add a test to assert the exception?

**Review by rui-mo — COMMENTED** [review]
Thanks.

File: velox/functions/sparksql/String.h
Comment (zhli1142015) [reply velox/functions/sparksql/String.h:1200]: Added, thanks.

File: velox/functions/sparksql/Register.cpp
Comment (jinchengchenghh) [inline velox/functions/sparksql/Register.cpp:463]: "string_repeat" -> prefix + "string_repeat"

File: velox/functions/sparksql/tests/StringTest.cpp
Comment (jinchengchenghh) [inline velox/functions/sparksql/tests/StringTest.cpp:455]: should also test string_repeat function name

File: velox/functions/sparksql/Register.cpp
Comment (zhli1142015) [reply velox/functions/sparksql/Register.cpp]: Good catch, thanks

File: velox/functions/sparksql/tests/StringTest.cpp
Comment (zhli1142015) [reply velox/functions/sparksql/tests/StringTest.cpp]: Added, thanks

**Review by jinchengchenghh — APPROVED** [review]

File: velox/docs/functions/spark/string.rst
Comment (mbasmanova) [inline velox/docs/functions/spark/string.rst:233]: nit: It would be nice to make 'repeat' a link using :spark:func:.

File: velox/docs/functions/spark/string.rst
Comment (mbasmanova) [inline velox/docs/functions/spark/string.rst:148]: typo: empry -> empty

File: velox/docs/functions/spark/string.rst
Comment (mbasmanova) [inline velox/docs/functions/spark/string.rst:147]: It would be nice to add a limit on the total size of the result string. Maybe, something like in lpad / rpad. Does Spark has a limit?

```
template <bool lpad, bool isAscii, typename TOutString, typename TInString>
FOLLY_ALWAYS_INLINE void pad(
    TOutString& output,
    const TInString& string,
    const int64_t size,
    const TInString& padString) {
  static constexpr size_t padMaxSize = 1024 * 1024; // 1MB
```

File: velox/docs/functions/spark/string.rst
Comment (mbasmanova) [reply velox/docs/functions/spark/string.rst]: @zhli1142015 You can reproduce fuzzer failures using --seed 2932607463. You can then debug these using standard tools.

Overall, it would be nice to add a limit on the size of the output string.

https://github.com/facebookincubator/velox/pull/10218#discussion_r1652969475

File: velox/docs/functions/spark/string.rst
Comment (zhli1142015) [reply velox/docs/functions/spark/string.rst:147]: > Does Spark has a limit?

I think in spark the limit is int32.max, I think it's a too big value to fuzzer test.
https://github.com/apache/spark/blob/a50b30d7a02bf45f1ddb8db0be6779b1441e1d4d/common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java#L870

Comment (mbasmanova) [reply velox/docs/functions/spark/string.rst:147]: > I think it's a too big value to fuzzer test.

More importantly this is too large for production. How about we add 1 MB limit to the size of the result string?

Comment (zhli1142015) [reply velox/docs/functions/spark/string.rst:147]: I'm ok for 1 MB limit, @rui-mo , is this ok to you?

Comment (rui-mo) [reply velox/docs/functions/spark/string.rst:147]: @zhli1142015 Yes, it makes sense to me.

Comment (zhli1142015) [reply velox/docs/functions/spark/string.rst]: Updated, thanks.

**Comment by zhli1142015** [comment]
@rui-mo , do you have more comments? Thanks.

File: velox/functions/sparksql/String.h
Comment (rui-mo) [inline velox/functions/sparksql/String.h:1185]: nit: add some comments for this function. Some example in the replace function:

> /// replace(input, replaced) -> varchar
///
///     Removes all instances of ``replaced`` from ``input``.
///     If ``replaced`` is an empty string, returns the original ``input``
///     string.

Comment (zhli1142015) [reply velox/functions/sparksql/String.h:1191]: added, thanks

File: velox/functions/sparksql/String.h
Comment (rui-mo) [inline velox/functions/sparksql/String.h:1187]: nit: 
> repeat(string, n)
string_repeat(string, n) -> varchar

> Returns the string which repeats the give ``string`` ``n`` times.

**Review by rui-mo — APPROVED** [review]

**Comment by PHILO-HE** [comment]
@zhli1142015, one minor comment. I note you say "repeat is used for other fucntions for Presto, so register this with a different name.". We have removed presto's scalar function registration from Gluten. There should be no conflict issue if we use the registered name "repeat" in Gluten. So we can remove the unnecessary alias name `string_repeat`.

**Comment by zhli1142015** [comment]
> @zhli1142015, one minor comment. I note you say "repeat is used for other fucntions for Presto, so register this with a different name.". We have removed presto's scalar function registration from Gluten. There should be no conflict issue if we use the registered name "repeat" in Gluten. So we can remove the unnecessary alias name `string_repeat`.

Thanks for sharing this, updated.

**Review by PHILO-HE — APPROVED** [review]
Looks nice!

**Pull request closed without merging**
